### PR TITLE
Prioritised tax calculation strategies

### PR DESCRIPTION
**Pull request by mckelviemark**
Added support for prioritised tax calculation strategies. This includes a global taxation setting for the default tax calculation strategy. This includes two strategies, one being an order items based strategy and the other being order item units based. This work should make it easy for client projects to create and drop in custom strategies.

### CODE REVIEW

File: src/Sylius/Bundle/CoreBundle/Resources/config/services.xml [hunk 1]
```diff
@@ -235,22 +238,62 @@
...
+        </service>
+
+        <service id="sylius.taxation.order_items_based_strategy" class="%sylius.taxation.composite_order_taxes_calculation_strategy.class%">
+            <argument key="type">order_items_based</argument>
+            <argument type="service">
+                <service
+                        factory-service="sylius.settings.manager"
+                        factory-method="loadSettings"
+                        class="Sylius\Bundle\SettingsBundle\Model\Settings"
+                >
```
Comment (peteward) [inline src/Sylius/Bundle/CoreBundle/Resources/config/services.xml]: I'm not sure about the multi-line split, seeing as they don't stick to 80 chars in their XML, but a small detail

Comment (peteward) [inline src/Sylius/Bundle/CoreBundle/Resources/config/services.xml]: NOTE: the settings bundle has a huge pending PR and a drastic change is imminent...

File: src/Sylius/Bundle/CoreBundle/Settings/TaxationSettingsSchema.php [hunk 2]
```diff
@@ -26,15 +28,13 @@
 class TaxationSettingsSchema implements SchemaInterface
 {
     /**
-     * Zone repository.
+     * Zone repository
```
Comment (peteward) [inline src/Sylius/Bundle/CoreBundle/Settings/TaxationSettingsSchema.php]: you can just remove this comment

File: src/Sylius/Bundle/CoreBundle/Settings/TaxationSettingsSchema.php [hunk 3]
```diff
@@ -26,15 +28,13 @@
...
      *
      * @var ObjectRepository
      */
-    private $zoneRepository;
+    protected $zoneRepository;
```
Comment (peteward) [inline src/Sylius/Bundle/CoreBundle/Settings/TaxationSettingsSchema.php]: why change this?

File: src/Sylius/Component/Core/Taxation/Applicator/CompositeOrderTaxesCalculationStrategy.php [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+namespace Sylius\Component\Core\Taxation\Applicator;
+
+use Sylius\Bundle\SettingsBundle\Model\Settings;
+use Sylius\Component\Addressing\Model\ZoneInterface;
+use Sylius\Component\Core\Model\OrderInterface;
+
+/**
+ * @author Mark McKelvie <mark.mckelvie@reiss.com>
+ */
+class CompositeOrderTaxesCalculationStrategy implements OrderTaxesApplicatorInterface
```
Comment (peteward) [inline src/Sylius/Component/Core/Taxation/Applicator/CompositeOrderTaxesCalculationStrategy.php]: should there be a common `TaxApplicatorInterface`? Is there any benefit or requirement for an `OrderTaxesApplicatorInterface` explicitly?

File: src/Sylius/Component/Core/Taxation/Applicator/CompositeOrderTaxesCalculationStrategy.php [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+     * @var OrderTaxesApplicatorInterface[]
+     */
+    protected $applicators;
+
+    /**
+     * @param string $type
+     * @param Settings $settings
+     * @param OrderTaxesApplicatorInterface[] $applicators
+     */
+    public function __construct($type, Settings $settings, array $applicators)
```
Comment (peteward) [inline src/Sylius/Component/Core/Taxation/Applicator/CompositeOrderTaxesCalculationStrategy.php]: does `$type` need to be injected? Isn't it dictated by the class? should this be a constant?

File: src/Sylius/Component/Core/Taxation/Applicator/OrderItemUnitsTaxesApplicator.php [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function apply(OrderInterface $order, ZoneInterface $zone)
+    {
+        foreach ($order->getItemUnits() as $unit) {
+            $item = $unit->getOrderItem();
+            $taxRate = $this->taxRateResolver->resolve($item->getVariant(), ['zone' => $zone]);
```
Comment (peteward) [inline src/Sylius/Component/Core/Taxation/Applicator/OrderItemUnitsTaxesApplicator.php]: hmmm....

It might be more efficient to have 2 loops, first loops around orderItems so you can resolve the tax rate just once per `OrderItem`, then an inner loop which calculates and applies it for each unit in the item?

File: src/Sylius/Component/Core/Taxation/Applicator/OrderItemUnitsTaxesApplicator.php [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    {
+        return 'order_item_units_only';
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function supports(OrderInterface $order, ZoneInterface $zone)
+    {
+        return false;
```
Comment (peteward) [inline src/Sylius/Component/Core/Taxation/Applicator/OrderItemUnitsTaxesApplicator.php]: huh?

File: src/Sylius/Component/Core/Taxation/Applicator/CompositeOrderTaxesCalculationStrategy.php (hunk 4 above)
Comment (peteward) [inline src/Sylius/Component/Core/Taxation/Applicator/CompositeOrderTaxesCalculationStrategy.php]: ok ignore me, this is at the strategy level... Should a Strategy implement an Applicator interface? hmmm

File: src/Sylius/Component/Core/Taxation/Applicator/OrderItemUnitsTaxesApplicator.php [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Component\Core\Taxation\Applicator;
```
Comment (peteward) [inline src/Sylius/Component/Core/Taxation/Applicator/OrderItemUnitsTaxesApplicator.php]: newline after this

File: src/Sylius/Component/Core/Taxation/Applicator/OrderTaxesApplicatorInterface.php [hunk 9]
```diff
@@ -9,21 +9,33 @@
...
     public function apply(OrderInterface $order, ZoneInterface $zone);
+
+    /**
+     * @return string
+     */
+    public function getType();
+
+    /**
+     * @param OrderInterface $order
+     * @param ZoneInterface  $zone
```
Comment (peteward) [inline src/Sylius/Component/Core/Taxation/Applicator/OrderTaxesApplicatorInterface.php]: sylius don't align comments

File: src/Sylius/Bundle/TaxationBundle/Form/Type/TaxCalculationStrategyChoiceType.php [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+use Sylius\Component\Taxation\Strategy\TaxCalculationStrategies;
+use Symfony\Component\Form\AbstractType;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+
+/**
+ * @author Mark McKelvie <mark.mckelvie@reiss.com>
+ */
+class TaxCalculationStrategyChoiceType extends AbstractType
+{
+    protected $strategies = [
```
Comment (peteward) [inline src/Sylius/Bundle/TaxationBundle/Form/Type/TaxCalculationStrategyChoiceType.php]: any reason to be protected instead of private?

**Comment by peteward** [comment]
Mark, some really, really excellent work here.

Why have you made the PR into `mm-tax-applicator` instead of `master`? What other changes would be included in the PR?  I think I'd like to see the whole picture.

I only have one significant design issue and that's the Composite/Strategy structure. My _feeling_ is that we should move away from the labelling that the Strategies are CompositeApplicators. They are definitely a kind of wrapping layer around Applicators, but they are more than a standard Composite pattern and I think can benefit from a clearer separation. This is my _suggestion_, but we should discuss whether you do this now or we look for Sylius feedback first.
- Don't make the strategies implement an `XXXXApplicatorInterface`, make it a `TaxStrategyInterface`
- Establish some separation between Applicators and Strategies by differing the methods such as `applyTaxes()` on the `Strategy` and `apply` on the `Applicator`.
- You have one achilles heel in overriding which is the use of some hard-coded strings and the `TaxCalculationStrategies` class.
  - Firstly, I don't think this class is needed, because if you want to add a new strategy you either have to override/extend this class or your strategy just won't be listed.
  - Secondly I don't think it offers anything anyway. It's used to identify which strategies should be available through settings, but shouldn't you use the registry here? Otherwise it's not going to be truely flexible because you would have to override the form and add your hard-coded strategy option. So if you instead inject the registry and use this to get the name of strategies you have a double-win.

UPDATE: I've just seen why you called the strategy `Composite` because it's reusable for either OrderItem or OrderItemUnit calculations - clever. But I still think we should remove the work Composite because it's not (quite) the same pattern.

Have a read through of this and let me know your thoughts. I think the changes I'm suggesting are quite small so maybe it's worth doing them first if you agree?

I would definitely like to see the full PR too.

File: src/Sylius/Component/Core/Taxation/Applicator/CompositeOrderTaxesCalculationStrategy.php (hunk 5 above)
Comment (peteward) [inline src/Sylius/Component/Core/Taxation/Applicator/CompositeOrderTaxesCalculationStrategy.php]: Ignore this, I see what you're doing here.

File: src/Sylius/Component/Core/Taxation/Applicator/OrderItemUnitsTaxesApplicator.php (hunk 7 above)
Comment (peteward) [inline src/Sylius/Component/Core/Taxation/Applicator/OrderItemUnitsTaxesApplicator.php]: OK again, I think perhaps a separation of the Strategy and Applicator interfaces means that you don't need `supports()` at the Applicator level - only Strategy.

File: src/Sylius/Bundle/CoreBundle/Settings/TaxationSettingsSchema.php (hunk 3 above)
Comment (mckelviemark) [inline src/Sylius/Bundle/CoreBundle/Settings/TaxationSettingsSchema.php]: I'll change it back. I was just trying to be consistent with the scope of their other member variables.

File: src/Sylius/Component/Core/Taxation/Applicator/OrderItemUnitsTaxesApplicator.php (hunk 7 above)
Comment (gonzalovilaseca) [inline src/Sylius/Component/Core/Taxation/Applicator/OrderItemUnitsTaxesApplicator.php]: +1

File: src/Sylius/Bundle/CoreBundle/Resources/config/services.xml (hunk 1 above)
Comment (mckelviemark) [inline src/Sylius/Bundle/CoreBundle/Resources/config/services.xml]: ok, as for the formatting of the multi-line split this is how they have formatted it elsewhere, so I'll keep this as is to be consistent.

File: src/Sylius/Bundle/TaxationBundle/Form/Type/TaxCalculationStrategyChoiceType.php (hunk 10 above)
Comment (mckelviemark) [inline src/Sylius/Bundle/TaxationBundle/Form/Type/TaxCalculationStrategyChoiceType.php]: It's protected because this is the same way that Sylius does it for other ChoiceTypes. When I saw a similar pattern elsewhere on Sylius I thought that to get it working I would list the strategy types here. Of course this means I would have to extend it on Reiss and we can avoid that by injecting it into the constructor instead. That's what I'm going to try now.

File: src/Sylius/Component/Core/Taxation/Applicator/OrderItemUnitsTaxesApplicator.php (hunk 6 above)
Comment (mckelviemark) [inline src/Sylius/Component/Core/Taxation/Applicator/OrderItemUnitsTaxesApplicator.php]: Good, I've updated this to calculate on every unit but to only resolve the tax rate once per item.

File: src/Sylius/Component/Core/Taxation/Applicator/OrderItemUnitsTaxesApplicator.php (hunk 8 above)
Comment (mckelviemark) [inline src/Sylius/Component/Core/Taxation/Applicator/OrderItemUnitsTaxesApplicator.php]: done.

File: src/Sylius/Component/Core/Taxation/Applicator/OrderTaxesApplicatorInterface.php (hunk 9 above)
Comment (mckelviemark) [inline src/Sylius/Component/Core/Taxation/Applicator/OrderTaxesApplicatorInterface.php]: done.

**Comment by mckelviemark** [comment]
@peteward The reason for the merge from mm-tax-applicator-review to mm-tax-applicator is that we already have a PR open with Sylius and everytime I push to that latter branch it triggers a build. Once I merge back to that branch we can add a PR comment to ping Sylius to review the approach. We'll need to point out that the PHPSpec is largely unfinished and possible Behat scenarios may need updating.

**Comment by peteward** [comment]
You could just close your PR

**Comment by mckelviemark** [comment]
You'll need to do that because it was you that opened it with my initial attempt. Anyway, we can merge this in very soon if you think it's fine for their review.

File: src/Sylius/Bundle/CoreBundle/Resources/config/services.xml [hunk 11]
```diff
@@ -235,22 +240,67 @@
...
+                    <argument>sylius_taxation</argument>
+                </service>
+            </argument>
+            <argument type="collection">
+                <argument type="service" id="sylius.taxation.order_items_taxes_applicator" />
+                <argument type="service" id="sylius.taxation.order_shipment_taxes_applicator" />
+            </argument>
+            <tag name="sylius.taxation.calculation_strategy" type="order_items_based" label="Order items based" />
+        </service>
```
Comment (peteward) [inline src/Sylius/Bundle/CoreBundle/Resources/config/services.xml:287]: newline here would be good

File: src/Sylius/Bundle/CurrencyBundle/Templating/Helper/CurrencyHelper.php [hunk 12]
```diff
@@ -58,13 +62,7 @@ public function convertAmount($amount, $currency = null)
...
 
     /**
-     * Convert amount and format it!
-     *
-     * @param int         $amount
-     * @param string|null $currency
-     * @param bool        $decimal
-     *
-     * @return string
+     * {@inheritdoc}
```
Comment (peteward) [inline src/Sylius/Bundle/CurrencyBundle/Templating/Helper/CurrencyHelper.php:65]: these changes look good but watch out for including unrelated changes in PRs, should be fine though as long as there's not many

**Comment by mckelviemark** [comment]
I think I should probably rename `TaxCalculationStrategyInterface` to just `TaxStrategyInterface`. This will still make sense and it will allow me to shorten the name of `UnsupportedTaxCalculationStrategyException` to `UnsupportedTaxStrategyException`.

**Comment by peteward** [comment]
I think it's fine as it is

**Comment by mckelviemark** [comment]
ok. :-)
